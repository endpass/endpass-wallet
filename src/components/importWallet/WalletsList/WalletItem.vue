<template>
  <div class="menu">
    <ul class="menu-list">
      <li>
        <a :class="{ 'is-active': isActive }" @click="handleClick">
          <account :key="address" :address="address" :size="10">
            <template slot="balance">
              <div v-if="isLoading" class="column has-spinner">
                <v-spinner :is-loading="isLoading" class="is-transparent" />
              </div>
              <balance
                v-else
                :amount="balance"
                :currency="activeCurrency.name"
                @update="getBalance"
              />
            </template>
          </account>
        </a>
      </li>
    </ul>
  </div>
</template>

<script>
import { mapState, mapActions, mapGetters } from 'vuex';
import Account from '@/components/Account';
import Balance from '@/components/Balance';

export default {
  name: 'WalletItem',
  props: {
    address: {
      type: String,
      required: true,
    },
    isActive: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    isLoading: true,
    balance: null,
  }),
  computed: {
    ...mapState('web3', ['activeCurrency']),
    ...mapGetters('web3', ['activeNetwork']),
  },
  watch: {
    activeNetwork() {
      this.getBalance();
    },
  },
  methods: {
    ...mapActions('accounts', ['getBalanceByAddress']),

    async getBalance() {
      this.isLoading = true;

      try {
        const { address } = this;
        const res = await this.getBalanceByAddress(address);

        this.balance = res.balance;
      } catch (e) {
        this.balance = '0';
      } finally {
        this.isLoading = false;
      }
    },
    handleClick() {
      this.$emit('click', this.address);
    },
  },
  mounted() {
    this.getBalance();
  },
  components: {
    Account,
    Balance,
  },
};
</script>
