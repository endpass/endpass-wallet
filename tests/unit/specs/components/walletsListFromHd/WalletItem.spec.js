import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import UIComponents from '@endpass/ui';

import WalletItem from '@/components/walletsListFromHd/WalletItem';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VeeValidate);
localVue.use(UIComponents);

jest.useFakeTimers();

describe('WalletItem', () => {
  let accountsModule;
  let wrapper;

  beforeEach(() => {
    accountsModule = {
      namespaced: true,
      actions: {
        getBalanceByAddress: jest.fn(),
      },
    };
    const storeOptions = {
      modules: {
        accounts: accountsModule,
        web3: {
          namespaced: true,
          state: {
            activeCurrency: { name: 'usd' },
          },
          getters: {
            activeNetwork: jest.fn(),
          },
        },
      },
    };
    const store = new Vuex.Store(storeOptions);
    wrapper = shallowMount(WalletItem, {
      localVue,
      store,
      propsData: {
        address: 'addr',
      },
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('WalletItem');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('getBalance', () => {
      it('should call vuex getBalance with correct arguments', async () => {
        expect.assertions(4);

        expect(wrapper.vm.isLoading).toBe(false);

        accountsModule.actions.getBalanceByAddress.mockResolvedValueOnce({
          balance: '10',
        });
        await wrapper.vm.getBalance();

        expect(wrapper.vm.balance).toEqual('10');
        expect(accountsModule.actions.getBalanceByAddress).toBeCalledWith(
          expect.anything(),
          'addr',
          undefined,
        );
        expect(wrapper.vm.isLoading).toBe(false);
      });

      it('should throw error', async () => {
        expect.assertions(4);

        const error = new Error('Balance error');

        accountsModule.actions.getBalanceByAddress.mockRejectedValue(error);
        expect(wrapper.vm.isLoading).toBe(false);

        await wrapper.vm.getBalance();

        expect(wrapper.vm.balance).toEqual('0');
        expect(accountsModule.actions.getBalanceByAddress).toBeCalledWith(
          expect.anything(),
          'addr',
          undefined,
        );
        expect(wrapper.vm.isLoading).toBe(false);
      });
    });
  });
});
