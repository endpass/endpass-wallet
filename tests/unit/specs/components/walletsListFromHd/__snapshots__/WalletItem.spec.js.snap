// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletItem render should render initial state of the component 1`] = `
<div
  class="menu"
>
  <ul
    class="menu-list"
  >
    <li>
      <a
        class=""
      >
        <account-stub
          address="addr"
          currency="ETH"
          size="10"
        >
          <template>
            <balance-stub
              amount="0"
              currency="usd"
              decimals="18"
              price="1"
              round="4"
            />
          </template>
           
          <!---->
        </account-stub>
      </a>
    </li>
  </ul>
</div>
`;
